### ISSUE TITLE
Dependency injection does not find startup class (Orleans V1.2.3)

### ISSUE THREAD

**Issue by peter-perot** [dotnet/orleans#2309]
I'm using Orleans since March 2015 in production. Those days we used an ugly service locator to resolve dependencies from our grains.

Since DI in Orleans has matured over time I now want to get rid of that service locator. Last week I have been playing around with ASP.NET Core DI and came to the conclusion that it will suffice our requirements (formerly we used Microsoft Unity DI).

I have put all my bootstrapping code into the startup class and tweaked my server-side XML configuration file as described here: https://dotnet.github.io/orleans/Documentation/Advanced-Concepts/Dependency-Injection.html

However: The `ConfigureServices` methods never gets called. Even if I put some intentionally misspelled full qualified class name into the `<Startup>` tag nothing happens (no error message or something like that).

Today I started a simple HelloWorld project from scratch using Orleans 1.2.3 (this is the version our production uses) with the console application silo host trying to get DI to work, but it still does not work: the bootstrapping code is not executed.

Since I have no clue what is missing I started a question at Stackoverflow (http://stackoverflow.com/questions/40046319/dependency-injection-with-orleans-version-1-2-3-not-working), but no answers until now.

Anyone any hint for me?

**Comment by peter-perot** [comment]
Updating to Orleans 1.3.0 did not solve the problem.

My config file:

```xml
<?xml version="1.0" encoding="utf-8"?>
<OrleansConfiguration xmlns="urn:orleans">
  <Globals>
    <Startup Type="Grains.Startup, Grains" />
    <StorageProviders>
      <Provider Type="Orleans.Storage.MemoryStorage" Name="MemoryStore" />
      <!-- To use Azure storage, uncomment one of the following lines: -->
      <!--<Provider Type="Orleans.Storage.AzureTableStorage" Name="AzureStore" DataConnectionString="UseDevelopmentStorage=true" />-->
      <!--<Provider Type="Orleans.Storage.AzureTableStorage" Name="AzureStore" DataConnectionString="DefaultEndpointsProtocol=https;AccountName=MYACCOUNTNAME;AccountKey=MYACCOUNTKEY" />-->
    </StorageProviders>
    <SeedNode Address="localhost" Port="22222" />
  </Globals>
  <Defaults>
    <Networking Address="localhost" Port="22222" />
    <ProxyingGateway Address="localhost" Port="40000" />
    <Tracing DefaultTraceLevel="Warning" TraceToConsole="true" TraceToFile="{0}-{1}.log">
      <TraceLevelOverride LogPrefix="Application" TraceLevel="Info" />
    </Tracing>
    <Statistics MetricsTableWriteInterval="30s" PerfCounterWriteInterval="30s" LogWriteInterval="300s" WriteLogStatisticsToTable="true"/>
  </Defaults>
  <Override Node="Primary">
    <Networking Address="localhost" Port="22222" />
    <ProxyingGateway Address="localhost" Port="40000" />
  </Override>
</OrleansConfiguration>
```

My startup class:

```csharp
using System;
using Microsoft.Extensions.DependencyInjection;

namespace Grains
{
    public class Startup
    {
        public IServiceProvider ConfigureServices(IServiceCollection services)
        {
            // no bootstrapping - just for testing purposes
            return services.BuildServiceProvider();
        }
    }
}
```

The assembly is named `Grains.dll`.

The only way I get this to work is programmatically:

```csharp
var config = ClusterConfiguration.LocalhostPrimarySilo();
config.UseStartupType<Startup>();
```

[HINT: Extension method `UseStartupType` does not exist in version 1.2.3.]

**Comment by jdom** [comment]
Hi @peter-perot, the `Startup` element should go into the `Defaults` node configuration, instead of the `Globals` XML element.
Additionally, you can manually set the node configuration programmatically even in 1.2.3, just take a look at the implementation of the [UseStartupType extension method](https://github.com/dotnet/orleans/blob/1.3.0/src/Orleans/Configuration/ConfigurationExtensions.cs#L66), which is just syntactic sugar.

**Comment by peter-perot** [comment]
See PR #2313.

**Comment by peter-perot** [comment]
@jdom: Thanks, putting the `Startup` tag into the `Defaults` section did the trick.

**Comment by peter-perot** [comment]
One more question: It works fine with Orleans 1.3.0. But when I create the same Hello-World project with Version 1.2.3, I get the following exception:

```
Could not load type 'Microsoft.Extensions.DependencyInjection.ServiceCollectionE
xtensions' from assembly 'Microsoft.Extensions.DependencyInjection.Abstractions,
 Version=1.0.0.0, Culture=neutral, PublicKeyToken=adb9793829ddae60'.
Exception = System.TypeLoadException: Could not load type 'Microsoft.Extensions.
DependencyInjection.ServiceCollectionExtensions' from assembly 'Microsoft.Extens
ions.DependencyInjection.Abstractions, Version=1.0.0.0, Culture=neutral, PublicK
eyToken=adb9793829ddae60'.
   at Orleans.Runtime.Startup.ConfigureStartupBuilder.RegisterSystemTypes()
   at Orleans.Runtime.Startup.ConfigureStartupBuilder.ConfigureStartup(String st
artupTypeName)
   at Orleans.Runtime.Silo..ctor(String name, SiloType siloType, ClusterConfigur
ation config, ILocalDataStore keyStore)
   at Orleans.Runtime.Silo..ctor(String name, SiloType siloType, ClusterConfigur
ation config)
   at Orleans.Runtime.Host.SiloHost.InitializeOrleansSilo()
```

I tried the same with our production (which also runs with version 1.2.3), and the same exception occurs.

Both assemblies - `Microsoft.Extensions.DependencyInjection.Abstractions.dll` v1.0.0.0 and `Microsoft.Extensions.DependencyInjection.dll` v1.0.0.0 - are correctly placed in the silo directory.

**Comment by sergeybykov** [comment]
Is this because of https://github.com/dotnet/orleans/issues/1890?

**Comment by jdom** [comment]
no, it was of incorrect documentation and lack of the Startup element in the XSD. Orleans 1.3.0 just had an extension method in its programmatic config that was easier to find, that's all

**Comment by attilah** [comment]
I think this issue for 1.2.3 is because the Microsoft.Orleans.Runtime package for 1.2.3 did not contain the reference to the DI packages, which is fixed in 1.3.0.

Check the dependencies:
1.3.0: https://www.nuget.org/packages/Microsoft.Orleans.OrleansRuntime/1.3.0
1.2.3: https://www.nuget.org/packages/Microsoft.Orleans.OrleansRuntime/1.2.3
